### PR TITLE
chore: creare Api stack 

### PR DESCRIPTION
**Pull request by Balou9**
add RestApi Distribution and Lambda Functions

### CODE REVIEW

File: storage/stack.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (chiefbiiko) [inline storage/stack.yml:1]: should go to `compute`

File: compute/openapi-injection.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
Comment (chiefbiiko) [inline compute/openapi-injection.sh:1]: if uncommented/unused let's not commit

File: compute/stack.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      - Label:
+          default: API
+        Parameters:
+          - FunctionMemorySize
+          . FunctionTimeOut
+
+Parameters:
+  FunctionMemorySize:
+    Description: how many megabytes of ram to allocate for the lambda
+    AllowedPattern: Number
```
Comment (chiefbiiko) [inline compute/stack.yml:16]: Type

File: compute/stack.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    Default: 128
+
+  FunctionTimeOut:
+    Description: Lambda timeout in seconds
+    AllowedPattern: Number
```
Comment (chiefbiiko) [inline compute/stack.yml:21]: Type

File: compute/stack.yml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    Default: 3
+
+Resources:
+  PayolaRestApi:
+    Type: AWS::ApiGateway::RestApi
+    Properties:
+      Name: PayolaRestApi
+      ApiKeySourceType: X-Api-Key
```
Comment (chiefbiiko) [inline compute/stack.yml:29]: this can either be `HEADER` or `AUTHORIZER` according to the [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-restapi.html)
but i don't think we even need it here
maybe drop it until we know we need this prop

File: compute/stack.yml [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      EndpointConfiguration: PRIVATE
```
Comment (chiefbiiko) [inline compute/stack.yml:30]: you can also drop this config as that is for setting up private apis but we certainly want a public one

File: compute/stack.yml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      # Body: # Flag serves as a placeholder for openapi.yml
+      #   OPENAPI
+
+  LambdaExecutionRole:
+  Type: AWS::IAM::Role
```
Comment (chiefbiiko) [inline compute/stack.yml:35]: indentation

File: compute/stack.yml [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  Properties:
+    AssumeRolePolicyDocument:
+      Version: 2012-10-17
+      Statement:
+      - Effect: Allow
+        Prinipal:
+          Service:
+            - lambda.amazonaws.com
+        Action:
+          - "sts:AssumeRole"
```
Comment (chiefbiiko) [inline compute/stack.yml:45]: the quotes are unnecessary, maybe just simplify to
`Action: sts:AssumeRole`

File: compute/stack.yml [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    Policies:
+      - PolicyName: LambdaExecutionPolicy
+        PolicyDocument:
+          Version: 2012-10-17
+          Statement:
+            - Effect: Allow
+              Action:
+                - dynamodb:GetItem
+                - dynamodb:PutItem
```
Comment (chiefbiiko) [inline compute/stack.yml:54]: here i suggest to change it to 
```
# TODO: narrow down permissions with a lambda exec role for every handler 
Action: dynamodb:*
```
for intial development
then narrow them down once we actually implement handlers

File: compute/stack.yml [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+  CreateClaimLambda:
+    Type: AWS::Lambda::Function
+    Properties:
+      Runtime: rust
```
Comment (chiefbiiko) [inline compute/stack.yml:59]: dont think its a valid runtime identifier
needs to be `provided`
and we bundle the runtime (an executable called `bootstrap`) in either the lambda zip archive or a lambda **layer** zip archive
bg: when a lambda container (a linux env) starts or *boots* by executing the executable located at `/opt/bootstrap`
for that we simply need to provide a `runtime.zip` as lambda layer that contains the `bootstrap` file at archive root

File: compute/stack.yml [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+      Handler: handler.createClaim
```
Comment (chiefbiiko) [inline compute/stack.yml:60]: cargo will supply us with a `lib.rs` and rust prefers snake_case
so this probly should be `Handler: lib.create_claim`

File: compute/stack.yml (hunk 10 above)
Comment (Balou9) [reply compute/stack.yml]: like this?

**Comment by Balou9** [comment]
@chiefbiiko : `CreateLambdaLayer` resource and lambda `Layers` and `Runtime` properties are ready for review. 
Also please note the todo proposal in line 122.

File: compute/stack.yml [hunk 12]
```diff
@@ -73,20 +95,56 @@ Resources:
 
           - Sid: AllowLambdaListBucket
             Effect: Allow
-            Principal:
+            Principal:BucketName
```
Comment (chiefbiiko) [inline compute/stack.yml]: typo

File: compute/stack.yml [hunk 13]
```diff
@@ -73,20 +95,56 @@ Resources:
...
+  Properties:
+    Tags:
+      - Key: "payola:name"
+        Value: !Sub ${AWS::StackName}-distribution
+      DistributionConfig:
+        DefaultCacheBehavior:
+          AllowedMethods:
+            - GET
+            - POST
+            - PUT
```
Comment (chiefbiiko) [inline compute/stack.yml:130]: dont allow write methods (PUT,POST) for the distribution - it doesn't make sense

irrelevant background:
besides GET, allowing HEAD and OPTIONS lets us respond to clients/browsers that check for CORS availability

File: compute/stack.yml [hunk 14]
```diff
@@ -110,71 +168,58 @@ Resources:
...
       MemorySize: !Ref FunctionMemorySize
       Timeout: !Ref FunctionTimeout
-      Layers:
-        - !Ref CreateLambdaLayer
 
   GetClaimStatusLambda:
     Type: AWS::Lambda::Function
     Properties:
-      Runtime: provided
+      Runtime: go1.x
```
Comment (chiefbiiko) [inline compute/stack.yml]: cfn lambda resource definitions require a `Code` property with the s3 location of our bundle - i think bucket_name and bundle_key should be params since we will feed those values in from some cd pipeline

File: index.html [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
```
Comment (chiefbiiko) [inline index.html:1]: this html can go to trash as well i think

File: storage/index.js [hunk 16]
```diff
@@ -0,0 +1,3 @@
+module.exports.getbalancehandler = function getbalancehandler (event) {
+  return "Hello World"
```
Comment (chiefbiiko) [inline storage/index.js:2]: not sure if just returning a string literal will work
usually, when using lambdas with apigateway (this mode is called apigw proxy integration), one has to comply to [this response format](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format)

File: compute/stack.yml [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+      ParentId: !Ref GasProviderResource
+      PathPart: balance
+      RestApiId: !Ref RestApi
+
+  BalanceResource:
+    Type: AWS::ApiGateway::Resource
+    Properties:
+      ParentId: !Ref BalanceResourceCollection
+      PathPart: "{balanceId}"
+      RestApiId: !Ref RestApi
```
Comment (chiefbiiko) [inline compute/stack.yml:197]: just thinking out loud
rite now we have'
`GET gas-provider/balance/{balanceId}`
here im not sure what `balanceId` is..
but here (below) the hierarchy seems odd..
`GET balance/gas-provider/{providerId}`
or maybe just
`GET balance`?
(`providerId` is kinda obolete since we will get a customer identifier in a cognito token)

File: compute/stack.yml [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+    Type: AWS::Lambda::Permission
+    DependsOn:
+      - RestApi
+      - GetBalanceHandler
+    Properties:
+      Action: lambda:InvokeFunction
+      FunctionName: !GetAtt GetBalanceHandler.Arn
+      Principal: apigateway.amazonaws.com
+      SourceArn: !Sub arn:aws:execute-api:${AWS:Region}:${AWS::AccountId}:${RestApi}/*
+
```
Comment (chiefbiiko) [inline compute/stack.yml:228]: the `SourceArn` value will need as many asterisks in the trailing part as their are path segments after the base url

so if we want 2 allow apigw access to a lambda from a method endpoint that looks kinda like here
`https://base_url/stage_name/parent_resource/child_resource`
here `base_url` corresponds 2 this part:
`arn:aws:execute-api:${AWS:Region}:${AWS::AccountId}:${RestApi}/`
we would need this
`!Sub arn:aws:execute-api:${AWS:Region}:${AWS::AccountId}:${RestApi}/*/*/*`

the tl;dr the asterisk only matches a siingle path segment (they do not match across slash`/`s)
